interop/src/clients/corecrypto/android.rs

```diff
@@ -0,0 +1,434 @@
+            }
+            Err(error) => {
+                panic!("Failed to launch application: {}", error)
+            }
+        }
+
+        log::info!("waiting for ready signal on system.out");
+
+        // Waiting for confirmation that the application has launched.
+        let mut line = String::new();
+        while !line.contains("Ready") {
+            line.clear();
+            output
+                .read_line(&mut line)
+                .expect("was expecting ready signal on stdout");
+        }
+
+        log::info!("application launched: {}", line);
+        Ok((process, output))
+    }
+
+    async fn execute(&self, action: String) -> Result<String> {
+        let args = [
+            "-s",
+            self.device.as_str(),
+            "shell",
+            "am",
+            "start",
+            "-W",
+            "-a",
+            "android.intent.action.RUN",
+            action.as_str(),
+        ];
+
+        log::info!("adb {}", args.join(" "));
+
+        Command::new("adb")
+            .args(args)
+            .output()
+            .expect("Failed to execute action");
+
+        let mut result = String::new();
+        let mut output = self.output.try_borrow_mut()?;
+
+        output.read_line(&mut result)?;
+
+        log::info!("{}", result);
+
+        let result: InteropResult = serde_json::from_str(result.trim())?;
+
+        match result {
+            InteropResult::Success { value } => Ok(value),
+            InteropResult::Failure { message } => Err(SimulatorDriverError { msg: message }.into()),
+        }
+    }
+}
+
+impl Drop for SimulatorDriver {
+    fn drop(&mut self) {
+        self.process.kill().expect("expected child process to be killed")
+    }
+}
+
+#[derive(Debug)]
+pub(crate) struct CoreCryptoAndroidClient {
+    driver: SimulatorDriver,
+    client_id: Vec<u8>,
+    #[cfg(feature = "proteus")]
+    prekey_last_id: Cell<u16>,
+}
+
+impl CoreCryptoAndroidClient {
+    pub(crate) async fn new() -> Result<Self> {
+        let client_id = uuid::Uuid::new_v4();
+        let client_id_str = client_id.as_hyphenated().to_string();
+        let client_id_base64 = general_purpose::STANDARD.encode(client_id_str.as_str());
+        let ciphersuite = CIPHERSUITE_IN_USE as u16;
+
+        let output = Command::new("adb")
+            .args(["get-serialno"])
+            .output()
+            .expect("Failed to get connected android device");
+
+        let device = String::from_utf8(output.stdout)
+            .expect("output is not valid utf8")
+            .trim()
+            .to_string();
+        let driver = SimulatorDriver::new(device, "com.wire.androidinterop".into());
+        log::info!("initialising core crypto with ciphersuite {}", ciphersuite);
+        driver
+            .execute(format!(
+                "--es action init-mls --es client_id {} --ei ciphersuite {}",
+                client_id_base64, ciphersuite
+            ))
```

Comment:
Nit: this could also be written as
```rust
.execute(format!(
 "--es action init-mls --es client_id {client_id_base64} --ei ciphersuite {ciphersuite}"
))
```